File: 108-convert_sorted_array_to_binary_search_tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None
        
        return self.helper(nums, 0, len(nums) - 1)

    def helper(self, nums, left, right):
        if left > right:
            return None
        
        # Always choose middle element as root
        mid = (left + right) // 2
        
        # Preorder traversal: node -> left -> right
        root = TreeNode(nums[mid])
        root.left = self.helper(nums, left, mid - 1)
        root.right = self.helper(nums, mid + 1, right)
        
        return root

# Example usage:
sol = Solution()
nums = [-10, -3, 0, 5, 9]
tree = sol.sortedArrayToBST(nums)

def preorder(node):
    if not node:
        return []
    return [node.val] + preorder(node.left) + preorder(node.right)

print(preorder(tree))  # Output: [0, -3, -10, 5, 9]

File: 108-convert_sorted_array_to_binary_search_tree.js
/* Given an integer array nums where the elements are sorted in ascending order, convert it to a height-balanced binary search tree.
 *
 * Example 1:
 *        0
 *    -3     9
 * -10     5
 * 
 * Input: nums = [-10,-3,0,5,9]
 * Explanation: [0,-10,5,null,-3,null,9] is also accepted:
 *        0
 *   -10     5
 *     -3      9
 * 
 * Example 2:
 *    3     1
 *  1         3
 * Input: nums = [1,3]
 * Output: [3,1]
 * Explanation: [1,null,3] and [3,1] are both height-balanced BSTs.
 * 
 * Constraints:
 * 1 <= nums.length <= 104
 * -104 <= nums[i] <= 104
 * Output: [0,-3,9,-10,null,5]
 * nums is sorted in a strictly increasing order.
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function (nums) {
    //base cases
    if (nums.length === 1) return new TreeNode(nums[0]);
    if (nums.length === 0) return null;

    //create a new TreeNode(center)
    let centerIdx = Math.floor(nums.length / 2);
    let root = new TreeNode(nums[centerIdx]);

    //set left node to center of left subtree
    let leftSubtree = nums.slice(0, centerIdx);
    root.left = sortedArrayToBST(leftSubtree);

    //set right node to center of right subtree
    let rightSubtree = nums.slice(centerIdx + 1, nums.length);
    root.right = sortedArrayToBST(rightSubtree);

    return root;
};
